### ISSUE TITLE
AutoMap self referencing class

### ISSUE THREAD

**Issue by bpaczkowski** [JoshClose/CsvHelper#974]
I have a class that references itself, is there a way to use it with AutoMapping? I'm getting StackOverflowExceptions if I try to AutoMap it now.

A second question, is there a way for header to always be written, regardless of whether there are any records or not?

**Comment by JoshClose** [comment]
Circular references should be fine. It should detect them and not follow. Can you give an example that reproduces this?

If you don't have any records but want to write the header, you can call `WriteHeader<T>` or `WriteHeader( Type type )`.

**Comment by bpaczkowski** [comment]
`WriteHeader` wasn't working for me before because I didn't use `Flush` on the underlying stream, and when I wrote some records it must have went over a threshold and it flushed automatically, so it seemed to me `WriteHeader` was only working when some records were written.

As to the circular reference, a simple example that causes the exception is a class A with a property of type A. I solved it simply by using a base class that does not include the circular reference.
